### PR TITLE
set type of statement when processing `GlobalRef`s

### PR DESCRIPTION
**Pull request by simeonschaub**
Because we previously didn't set the type when processing `GlobalRef`,
the optimizer was inserting unnecessary `PiNode`s.

Before:

```julia
julia> x::Int = 1
1

julia> function f()
           global x = 0
           while x<10
               x += 1
           end
           x
       end
f (generic function with 1 method)

julia> @code_typed f()
CodeInfo(
1 ──       (Main.x = 0)::Any
2 ┄─ %2  = Main.x::Any
│    %3  = (isa)(%2, Int64)::Bool
└───       goto #4 if not %3
3 ── %5  = π (%2, Int64)
│    %6  = Base.slt_int(%5, 10)::Bool
└───       goto #5
4 ── %8  = (%2 < 10)::Bool
└───       goto #5
5 ┄─ %10 = φ (#3 => %6, #4 => %8)::Bool
└───       goto #10 if not %10
6 ── %12 = Main.x::Any
│    %13 = (isa)(%12, Int64)::Bool
└───       goto #8 if not %13
7 ── %15 = π (%12, Int64)
│    %16 = Base.add_int(%15, 1)::Int64
└───       goto #9
8 ── %18 = (%12 + 1)::Int64
└───       goto #9
9 ┄─ %20 = φ (#7 => %16, #8 => %18)::Int64
│          (Main.x = %20)::Any
└───       goto #2
10 ─ %23 = Main.x::Any
└───       return %23
) => Int64
```

This PR:

```julia
julia> @code_typed f()
CodeInfo(
1 ─      (Main.x = 0)::Any
2 ┄ %2 = Main.x::Int64
│   %3 = Base.slt_int(%2, 10)::Bool
└──      goto #4 if not %3
3 ─ %5 = Main.x::Int64
│   %6 = Base.add_int(%5, 1)::Int64
│        (Main.x = %6)::Any
└──      goto #2
4 ─ %9 = Main.x::Int64
└──      return %9
) => Int64
```

### CODE REVIEW

File: test/compiler/inline.jl [hunk 1]
```diff
@@ -1069,3 +1069,18 @@ end
...
+        global x::Int = 0
+        function f()
+            global x = 0
+            while x<10
+                x += 1
+            end
+            x
+        end
+    end
+    src = code_typed1(f)
```
Comment (aviatesk) [inline test/compiler/inline.jl:1084]:
```suggestion
    src = code_typed1(m.f)
```

Comment (simeonschaub) [reply test/compiler/inline.jl]: Oops! That happens if you only test in the REPL... :smile:

**Comment by aviatesk** [comment]
Inference understands `+` is resolved to `::Int + ::Int`, so the inliner shouldn't insert that `PiNode`. On master, I got:
```julia
julia> x::Int = 1
1

julia> function f()
           global x = 0
           while x<10
               x += 1
           end
           x
       end
f (generic function with 1 method)

julia> @code_typed f()
CodeInfo(
1 ─      (Main.x = 0)::Any
2 ┄ %2 = Main.x::Any
│   %3 = Base.slt_int(%2, 10)::Bool
└──      goto #4 if not %3
3 ─ %5 = Main.x::Any
│   %6 = Base.add_int(%5, 1)::Int64
│        (Main.x = %6)::Any
└──      goto #2
4 ─ %9 = Main.x::Any
└──      return %9
) => Int64
```
so that test case should pass already.

Though we should set types for those `GlobalRef` statements.

**Comment by simeonschaub** [comment]
Huh, so this must have improved very recently? But you still think this change is a good idea?

**Comment by simeonschaub** [comment]
Yeah, I compared this to bfc94311badd3be87b49ef446473a058235fa19a before, so apparently it must have been fixed in the last 3 days.

**Comment by aviatesk** [comment]
Interestingly, #44095 has improved the situation -- previously the inliner computes call signature by itself (so can be affected by this change) but now it just trusts the result of inference.

**Review by aviatesk — APPROVED** [review]
I couldn't come up with a test case that demonstrates what is improved by this change, but we may want to use this type information in some new optimization pass in the future.

**Comment by aviatesk** [comment]
Is the failure on macOS buildbot a known issue? I rebuilt it multiple times, but it seems to persist.

**Review by JeffBezanson — APPROVED** [review]

File: test/compiler/inline.jl [hunk 2]
```diff
@@ -1069,3 +1069,18 @@ end
...
+            global x = 0
+            while x<10
+                x += 1
+            end
+            x
+        end
+    end
+    src = code_typed1(m.f)
+    @test count(x -> isa(x, Core.PiNode), src.code) == 0
+end
```
Comment (vtjnash) [inline test/compiler/inline.jl:1086]: There should not be a new module created in tests, unless essential (and other style nits)
```suggestion
global x44200::Int = 0
function f44200()
    global x = 0
    while x < 10
        x += 1
    end
    x
end
let src = code_typed1(m.f)
    @test count(x -> isa(x, Core.PiNode), src.code) == 0
end
```

File: test/compiler/inline.jl [hunk 3]
```diff
@@ -1069,3 +1069,15 @@ end
...
+
+global x44200::Int = 0
+function f44200()
+    global x = 0
+    while x < 10
+        x += 1
+    end
+    x
+end
+let src = code_typed1(m.f)
```
Comment (simeonschaub) [inline test/compiler/inline.jl:1081]:
```suggestion
let src = code_typed1(f44200)
```

File: test/compiler/inline.jl [hunk 4]
```diff
@@ -1069,3 +1069,15 @@ end
...
     issue41694(2)
 end
+
+global x44200::Int = 0
+function f44200()
+    global x = 0
+    while x < 10
+        x += 1
+    end
+    x
```
Comment (vtjnash) [inline test/compiler/inline.jl:1079]:
```suggestion
    global x44200 = 0
    while x44200 < 10
        x44200 += 1
    end
    x44200
```

**Comment by vtjnash** [comment]
FWIW, I don't see these PiNodes on my builds, so the test does not seem to do anything for me. Could be I am looking at the wrong thing though.

**Comment by simeonschaub** [comment]
Yes, as discussed above, https://github.com/JuliaLang/julia/pull/44095 actually already fixed this particular test case, but I think this is worthwhile regardless.

**Pull request merged**